exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Table.kt

```diff
@@ -246,6 +246,23 @@ class Join(
             table.columns.toMutableList()
         ) { it.joinPart.columns }
 
+    override val fields: List<Expression<*>>
+        get() = joinParts.flatMapTo(
+            table.fields.toMutableList()
+        ) {
+            when {
+                it.joinPart is QueryAlias ->
+                    it.joinPart.fields.map { field ->
+                        when (field) {
+                            is AliasOnlyExpression<*> ->
+                                field.alias("${it.joinPart.alias}.${field.origin.alias}").aliasOnlyExpression()
+                            else -> field
+                        }
```

Comment:
If `QueryAlias` is the problem, I'd consider using the logic we already have for it instead of creating a new `AliasOnlyExpression`.
The logic I'm referring to can be found in `QueryAlias.get()` and could be reused there instead of in `Join`, something like this:
```kt
// new prop in QueryAlias
internal val aliasedFields: List<Expression<*>> = query.set.fields.map { expression ->
    (expression as? Column<*>)?.clone()
        ?: (expression as? ExpressionAlias<*>)?.let {
            it.delegate.alias("$alias.${it.alias}").aliasOnlyExpression() // this could be a private fun
        }
        ?: expression
}

// then used in Join
override val fields: List<Expression<*>>
    get() = joinParts.flatMapTo(table.fields.toMutableList()) {
        when (it.joinPart) {
            is QueryAlias -> it.joinPart.aliasedFields
            else -> it.joinPart.fields
        }
    }
```